From be160af778f37bf534b6e5611ba3b10ad03a2947 Mon Sep 17 00:00:00 2001
From: Kevin Tung <kevin.tung.openbmc@gmail.com>
Date: Mon, 5 Oct 2026 14:46:58 +0800
Subject: [PATCH] meta-facebook: yosemite5: add APML recovery

An unexpected BMC reset during an SBI transaction can leave the target
device controller in a hung state or leave stale data in the RX/TX
FIFOs, which can corrupt data associated with subsequent SBI requests.

Execute the SB-RMI and SB-TSI reset recovery flows during BMC boot to
clear stale state in the SBI targets and restore APML communication.

Change-Id: I362925d3e97f781c2b895dfe7e78ac036591df5c
Signed-off-by: Kevin Tung <kevin.tung.openbmc@gmail.com>
---
 .../plat-svc/files/yosemite5-early-sys-init           | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/yosemite5-early-sys-init b/yosemite5-early-sys-init
index ccefa3b0f5..bfe0224c76 100644
--- a/yosemite5-early-sys-init
+++ b/yosemite5-early-sys-init
@@ -403,4 +403,15 @@ set_gpio FM_SMB_AUTH_MUX_OE_R_N 1
 set_gpio FM_SCM_LED_R_N 1
 set_gpio SGPIO_BMC_READY 1
 
+# An unexpected reset of BMC in the middle of SBI transactions can leave the
+# target device controllers in a hung state or RX/TX FIFOs in a non-empty
+# state that could skew the data bytes associated with SBI requests.
+# Execute SB-RMI and SB-TSI reset recovery flows to clear off stale state
+# in the SBI targets.
+if [ "$(get_gpio PWRGD_CPU_PWROK_1)" -eq 1 ]; then
+    echo "Host state is Running, recovering APML..."
+    apml_tool 0 --apml_recovery 0
+    apml_tool 0 --apml_recovery 1
+fi
+
 exit 0
-- 
2.47.3

